### ISSUE TITLE
When using auto_delta_encoding_order with certain chunk values, an out-of-bounds crash can occur

### ISSUE THREAD

**Issue by baz** [mwlon/pcodec#143]
I've created a small sample project which demonstrates the crash here: https://github.com/baz/pco-issue/blob/main/src/main.rs

If I specify the `delta_encoding_order`, then I can avoid this. I've only observed it to happen with the values in that project. I don't know what it is about the pattern there which causes the crash, i.e. `vec![167772161, 150994945, 167772161]`.

```
thread 'main' panicked at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/ans/encoding.rs:152:16:
index out of bounds: the len is 2 but the index is 2
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:597:5
   1: core::panicking::panic_fmt
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/panicking.rs:72:14
   2: core::panicking::panic_bounds_check
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/panicking.rs:180:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/slice/index.rs:261:10
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/slice/index.rs:18:9
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/vec/mod.rs:2728:9
   6: pco::ans::encoding::quantize_weights_to
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/ans/encoding.rs:152:16
   7: pco::ans::encoding::quantize_weights
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/ans/encoding.rs:176:21
   8: pco::wrapped::chunk_compressor::quantize_weights
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/wrapped/chunk_compressor.rs:122:33
   9: pco::wrapped::chunk_compressor::train_infos
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/wrapped/chunk_compressor.rs:175:22
  10: pco::wrapped::chunk_compressor::unsigned_new_w_delta_order
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/wrapped/chunk_compressor.rs:457:19
  11: pco::wrapped::chunk_compressor::choose_delta_encoding_order
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/wrapped/chunk_compressor.rs:543:21
  12: pco::wrapped::chunk_compressor::unsigned_new
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/wrapped/chunk_compressor.rs:585:5
  13: pco::wrapped::chunk_compressor::new
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/wrapped/chunk_compressor.rs:609:3
  14: pco::wrapped::file_compressor::FileCompressor::chunk_compressor
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/wrapped/file_compressor.rs:74:5
  15: pco::standalone::compressor::FileCompressor::chunk_compressor
             at /Users/baz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pco-0.1.2/src/standalone/compressor.rs:92:14
  16: pco_issue::main
             at ./src/main.rs:12:28
  17: core::ops::function::FnOnce::call_once
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/ops/function.rs:250:5
```

**Comment by mwlon** [comment]
Thanks for reporting this. It was a simple bug with a simple fix: https://github.com/mwlon/pcodec/pull/144

This would happen whenever there were an equal number `b` of bins and numbers (or deltas) with `b > 1`, which was possible for very small chunks.

**Comment by baz** [comment]
Thanks Martin

**Comment by mwlon** [comment]
The fix is released in 0.1.3 now.
